cloudinit/user_data.py

```diff
@@ -78,66 +85,96 @@ def __init__(self, paths):
+            )
         return accumulating_msg
 
-    def _process_msg(self, base_msg, append_msg):
+    def _process_msg(self, base_msg: Message, append_msg, require_pgp=False):
         def find_ctype(payload):
             return handlers.type_from_starts_with(payload)
 
         for part in base_msg.walk():
             if is_skippable(part):
                 continue
 
-            ctype = None
-            ctype_orig = part.get_content_type()
             payload = util.fully_decoded_payload(part)
-            was_compressed = False
 
-            # When the message states it is of a gzipped content type ensure
-            # that we attempt to decode said payload so that the decompressed
-            # data can be examined (instead of the compressed data).
-            if ctype_orig in DECOMP_TYPES:
-                try:
-                    payload = util.decomp_gzip(payload, quiet=False)
-                    # At this point we don't know what the content-type is
-                    # since we just decompressed it.
-                    ctype_orig = None
-                    was_compressed = True
-                except util.DecompressionError as e:
-                    error_message = (
-                        "Failed decompressing payload from {} of"
-                        " length {} due to: {}".format(
-                            ctype_orig, len(payload), e
-                        )
-                    )
-                    _handle_error(error_message, e)
-                    continue
+            ctype = part.get_content_type()
 
-            # Attempt to figure out the payloads content-type
-            if not ctype_orig:
-                ctype_orig = UNDEF_TYPE
             # There are known cases where mime-type text/x-shellscript included
             # non shell-script content that was user-data instead.  It is safe
             # to check the true MIME type for x-shellscript type since all
             # shellscript payloads must have a #! header.  The other MIME types
             # that cloud-init supports do not have the same guarantee.
-            if ctype_orig in TYPE_NEEDED + ["text/x-shellscript"]:
-                ctype = find_ctype(payload)
-            if ctype is None:
-                ctype = ctype_orig
+            if ctype in TYPE_NEEDED + ["text/x-shellscript"]:
+                ctype = find_ctype(payload) or ctype
+
+            if require_pgp and ctype != ENCRYPT_TYPE:
+                error_message = (
+                    "'require_pgp' was set true in cloud-init's base "
+                    f"configuration, but content type is {ctype}."
+                )
+                raise RuntimeError(error_message)
+
+            was_transformed = False
+
+            # When the message states it is transformed ensure
+            # that we attempt to decode said payload so that the transformed
+            # data can be examined.
+            parent_ctype = None
+            if ctype in TRANSFORM_TYPES:
```

Comment:
I made a necessary refactor to this section involving setting the mime type of transformed messages. Please double check my logic :slightly_smiling_face:  .

I did ensure that gzipped user data still works properly with this code.